refactor: migrated FEATURES dict settings to top-level in core files and fixed related test files.

Akanshu-2u opened the pull request:
### Description:
Refactored `FEATURES` settings to top-level in core files and updated related tests for consistency and maintainability.
### Solution:

- Modified the `SettingDictToggle` class to `SettingToggle`.
- Moved `FEATURES` setting to top level in core file.
- Fixed the related settings in test files leaving other unrelated `FEATURE` references intact. 

**Note:**  The unrelated `FEATURE` references are working fine due to the use of `featureproxy` , hence for the time being left those unchanged.

### Private JIRA Link:
[BOMS-200](https://2u-internal.atlassian.net/browse/BOMS-200)

### Reference PR:
https://github.com/openedx/edx-platform/pull/37067

### Note:

- On investigating how the top-level-settings are working in edx platform , found these few settings are still getting derived from the `FEATURES` dict. 
- Planned to move these settings to top level.

Akanshu-2u committed:
refactor: moved remaining feature dicts settings into top-level settings.

Akanshu-2u committed:
refactor: moved remaining feature dicts settings into top-level settings.

feanil reviewed (commented):
Thanks for the PR @Akanshu-2u, it looks like the right changes but maybe not enough fixes in the tests which may be trying to override the features dict still. Once the tests get fixed up, I think this would be fine to land.  Let me know when you've got the tests passing. Until then, I'm gonna mark this as a draft PR.

Akanshu-2u committed:
fix: fixed the test files

Akanshu-2u committed:
fix: fixed tehe pylint errors

Akanshu-2u committed:
fix: fixation of the cms ci failure

Akanshu-2u committed:
fix: fixed remaining feature settings for cms

Akanshu-2u committed:
fix: added fix for requirements

Akanshu-2u committed:
fix: added fix for lms tests

Akanshu-2u committed:
fix: resolved the test views issue

Akanshu-2u committed:
fix: configured views file and test_views

Akanshu-2u committed:
fix: fixed lint errors and assertion issues

Akanshu-2u committed:
fix: added fix for base url issue in test view

Akanshu-2u committed:
fix: added fix for base_url and assertion issue

Akanshu-2u committed:
fix: added configurations for base utl fix

Akanshu-2u committed:
fix: handled none issue for mfe config

Akanshu-2u committed:
fix: corrected override settings in test views

Akanshu-2u committed:
fix: added getattr defensive technique for view settings

Akanshu-2u committed:
fix: reverted views and test_views file

Akanshu-2u committed:
fix: added settings in views file

Akanshu-2u committed:
fix: added with patch within functions in test view

Akanshu-2u committed:
fix: rearranged the features in default_legacy_config

Akanshu-2u committed:
fix: fixing the tests  with clearing cache

Akanshu-2u committed:
fix: reverted test views to verify the CI check

Akanshu-2u committed:
fix: added cache clear in mfe config test

jcapphelix commented on cms/djangoapps/contentstore/tests/test_course_settings.py:
```diff
@@ -394,15 +395,15 @@ def test_invalid_pre_requisite_course(self):
         """
         with patch.dict("django.conf.settings.FEATURES", {
-            'ENTRANCE_EXAMS': feature_flags[0],
             'ENABLE_PUBLISHER': feature_flags[1]
```
If we are removing feature_flags, shouldn't `feature_flags[1]` be also removed ?

feanil replied:
I'm curious about this as well? Is it just that this is migrating some of the features but not everything?

Akanshu-2u replied:
Yes, we are migrating some of the features specific to this PR only, as with use of `featuresproxy` the other unrelated Feature settings are working well.

jcapphelix commented on cms/djangoapps/contentstore/tests/test_course_settings.py:
```diff
@@ -411,14 +412,15 @@ def test_visibility_of_entrance_exam_section(self, feature_flags):
 
         with mock.patch.dict('django.conf.settings.FEATURES', {
             'ENABLE_PUBLISHER': True,
             'ENABLE_MKTG_SITE': True,
-            'ENTRANCE_EXAMS': False,
-            'ENABLE_PREREQUISITE_COURSES': False
+            'ENABLE_PREREQUISITE_COURSES': False,
```
Are these supposed to be left as is ? Like within a feature flag ? 

Or is it something that is supposed to be removed later ?

jcapphelix commented on cms/djangoapps/contentstore/tests/test_utils.py:
```diff
@@ -934,7 +934,7 @@ def test_update_course_details_instructor_paced(self, mock_update):
         }
 
-        utils.update_course_details(mock_request, self.course.id, payload, None)
+        utils.update_course_details(mock_request, self.course.id, payload, self.course)
```
why are we changing the args ? This is un-related to features. 

Was it because of tests failing ?

Akanshu-2u replied:
Yes, the CI test was failing with this error `AttributeError: 'NoneType' object has no attribute 'entrance_exam_enabled'` . So changed the arg to fix this.

feanil replied:
I believe this is happening because the setting value for `ENTRANCE_EXAMS` is now being set to true instead of false which is not the correct expected behavior.  So I think the fix needs to be to patch the toggle rather than the item in the features dictionary not to change the test code here.

jcapphelix commented on common/djangoapps/util/tests/test_milestones_helpers.py:
```diff
@@ -61,8 +61,7 @@ def test_pre_requisite_courses_enabled(self, feature_flags):
 
         with patch.dict("django.conf.settings.FEATURES", {
```
We still have feature flags here ? Is it something that's supposed to be removed later ?

jcapphelix reviewed (changes requested):
While core file changes are all good. 

I have some questions on test cases, mainly :- 

- Why are some feature flags not removed ? 
- Is it something that is supposed to be removed later ?

Akanshu-2u commented:
> While core file changes are all good.
> 
> I have some questions on test cases, mainly :-
> 
> * Why are some feature flags not removed ?
> * Is it something that is supposed to be removed later ?

- The settings which where removed from `features` dict in the core files. Only those are modified in the test files leaving the other feature settings in place as due to `featureproxy` those will not create any issue. 
- Secondly, some args and cache changes are mainly due to the failure of the CI checks so have included those to prevent anything from breaking.
- The remaining feature instances in the test files can be removed, waiting on confirmation.

jcapphelix reviewed (approved):
My Queries are all answered. 

Good to go from my side.

kdmccormick reviewed (approved):
Thanks @Akanshu-2u , these changes all look good.

But, I am confused why you chose to update these references in particular. There are many remaining references to `FEATURES` that we will need to clean up in a future PR.

Were these particular references causing bugs?

Akanshu-2u commented:
> Thanks @Akanshu-2u , these changes all look good.
> 
> But, I am confused why you chose to update these references in particular. There are many remaining references to `FEATURES` that we will need to clean up in a future PR.
> 
> Were these particular references causing bugs?

I focused on fixing only the test files that referenced the `FEATURES` dictionary settings, which had been removed from the core files within this PR.
The settings were causing bugs resulting in test failures so was essential to update those test cases with override-settings.

kdmccormick commented:
> which had been removed from the core files within this PR.

Why did you choose to refactor those settings within this PR?

Akanshu-2u commented:
> > which had been removed from the core files within this PR.
> 
> Why did you choose to refactor those settings within this PR?

I guess this may help you with the explanation: https://github.com/openedx/edx-platform/issues/37396#issuecomment-3386024654

kdmccormick commented:
Okay, so you have chosen some remaining references, but not all of them, correct? That is fine, but the description of this PR says that it "moved remaining features dicts settings" and "Removed FEATURES dictionary". These are not true, and it is misleading to have them in the PR description when there are still so many FEATURES references remaining in the platform. Could you please edit your PR and commit message?

Akanshu-2u commented:
> Okay, so you have chosen some remaining references, but not all of them, correct? That is fine, but the description of this PR says that it "moved remaining features dicts settings" and "Removed FEATURES dictionary". These are not true, and it is misleading to have them in the PR description when there are still so many FEATURES references remaining in the platform. Could you please edit your PR and commit message?

Addressed. Modified the PR title and description which shall be more convenient for the changes achieved in this PR.

robrap commented:
I'm not clear if fixing a subset at a time adds some risk or not. Do we need to ensure changes are consistent for any particular setting, or does the proxy make that moot? Possibly not if someone had the FEATURES dict settings duplicated _and_ out-of-sync, but that might just move from one bad result to a different bad result.

In any case, if no one has a problem with it, can we delay landing this until after the ulmo branch has been cut? After that, there is less risk for 2U, and since this is a 2U development effort, that would be my preference. That said, if this did cause any issues, the known active early deployer would still want to know, so we could still announce in #risky-changes when we decide to merge.

Akanshu-2u committed:
Merge branch 'openedx:master' into BOMS-200-handle-features-setting-dict-depr

feanil commented on cms/djangoapps/contentstore/tests/test_course_settings.py:
```diff
@@ -388,15 +389,15 @@ def test_invalid_pre_requisite_course(self):
     )
     @override_waffle_flag(toggles.LEGACY_STUDIO_SCHEDULE_DETAILS, True)
-    def test_visibility_of_entrance_exam_section(self, feature_flags):
+    @patch.object(milestones_helpers.ENABLE_MILESTONES_APP, 'is_enabled', return_value=False)
```
Why patch the toggle rather than just using `override_setting` which is more standard and what you're using in some of the later files?

Akanshu-2u replied:
Implemented the suggestion.

feanil commented on cms/djangoapps/contentstore/views/tests/test_library.py:
```diff
@@ -161,8 +162,8 @@ def test_library_creator_status_with_no_course_creator_role_and_disabled_nonstaf
 
-    @mock.patch.dict('django.conf.settings.FEATURES', {'ENABLE_CONTENT_LIBRARIES': False})
-    def test_with_libraries_disabled(self):
+    @patch.object(toggles.ENABLE_CONTENT_LIBRARIES, 'is_enabled', return_value=False)
```
I think it's much clearer if you just override the django settings rather than the toggle.

Akanshu-2u replied:
Yes, following the suggestion I added override setting instead of patch toggle.

feanil commented on lms/djangoapps/mfe_config_api/tests/test_views.py:
```diff
@@ -30,6 +31,7 @@ class MFEConfigTestCase(APITestCase):
 
     def setUp(self):
         self.mfe_config_api_url = reverse("mfe_config_api:config")
+        cache.clear()
```
Why was this needed?

Akanshu-2u replied:
This change was needed because CI checks were repeatedly failing due to cached data being retained for the test file. Clearing the cache ensures the tests run cleanly and pass as expected.

feanil reviewed (commented):
Mostly looks good, just had a couple of questions.

Akanshu-2u committed:
Merge branch 'openedx:master' into BOMS-200-handle-features-setting-dict-depr

Akanshu-2u committed:
fix: fixed the patch toggles to override settings

Akanshu-2u committed:
fix: fixed the lint errors

feanil commented on cms/djangoapps/contentstore/tests/test_course_settings.py:
```diff
@@ -1128,16 +1129,16 @@ def test_fetch_initial_fields(self):
 
-    @patch.dict(settings.FEATURES, {'ENABLE_EXPORT_GIT': True})
-    def test_fetch_giturl_present(self):
+    @patch.object(toggles.EXPORT_GIT, 'is_enabled', return_value=True)
```
Is there a reason this isn't using `override_settings` as well?

Akanshu-2u replied:
Changed this to override settings and ensured other files do not require changes for override settings.

feanil reviewed (commented):
Looks like one more file needs to be updated and then this is good to land I think.

Akanshu-2u committed:
fix: changed patch toggle to override settings

Akanshu-2u committed:
Merge branch 'openedx:master' into BOMS-200-handle-features-setting-dict-depr

feanil reviewed (approved):
Changes look great, thanks @Akanshu-2u
